pkg/controller/cronjob_trigger_controller.go

```diff
@@ -0,0 +1,364 @@
+		c.logger.Errorf("Error processing %s (will retry): %v", key, err)
+		c.queue.AddRateLimited(key)
+	} else {
+		// err != nil and too many retries
+		c.logger.Errorf("Error processing %s (giving up): %v", key, err)
+		c.queue.Forget(key)
+		utilruntime.HandleError(err)
+	}
+
+	return true
+}
+
+func (c *CronJobTriggerController) syncCronJobTrigger(key string) error {
+	c.logger.Infof("Processing update to CronJob Trigger: %s", key)
+
+	ns, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	obj, exists, err := c.cronJobInformer.Informer().GetIndexer().GetByKey(key)
+	if err != nil {
+		return fmt.Errorf("Error fetching object with key %s from store: %v", key, err)
+	}
+
+	// this is an update when CronJob trigger API object is actually deleted, we dont need to process anything here
+	if !exists {
+		c.logger.Infof("Cronjob Trigger %s not found, ignoring", key)
+		return nil
+	}
+
+	cronJobtriggerObj := obj.(*kubelessApi.CronJobTrigger)
+
+	// CronJob trigger API object is marked for deletion (DeletionTimestamp != nil), so lets process the delete update
+	if cronJobtriggerObj.ObjectMeta.DeletionTimestamp != nil {
+
+		// If finalizer is removed, then we already processed the delete update, so just return
+		if !c.cronJobTriggerObjHasFinalizer(cronJobtriggerObj) {
+			return nil
+		}
+
+		// CronJob Trigger object should be deleted, so remove associated cronjob and remove the finalizer
+		_, err := c.clientset.BatchV2alpha1().CronJobs(ns).Get(fmt.Sprintf("trigger-%s", name), metav1.GetOptions{})
+		if err == nil {
+			err = c.clientset.BatchV2alpha1().CronJobs(ns).Delete(fmt.Sprintf("trigger-%s", name), &metav1.DeleteOptions{})
+			if err != nil && !k8sErrors.IsNotFound(err) {
+				c.logger.Errorf("Failed to remove CronJob created for CronJobTrigger Obj: %s due to: %v: ", key, err)
+				return err
+			}
+		}
+
+		// remove finalizer from the cronjob trigger object, so that we dont have to process any further and object can be deleted
+		err = c.cronJobTriggerObjRemoveFinalizer(cronJobtriggerObj)
+		if err != nil {
+			c.logger.Errorf("Failed to remove CronJob trigger controller as finalizer to CronJob Obj: %s due to: %v: ", key, err)
+			return err
+		}
+
+		c.logger.Infof("Cronjob trigger object %s has been successfully processed and marked for deletion", key)
+		return nil
+	}
+
+	// If CronJob trigger API in not marked with self as finalizer, then add the finalizer
+	if !c.cronJobTriggerObjHasFinalizer(cronJobtriggerObj) {
+		err = c.cronJobTriggerObjAddFinalizer(cronJobtriggerObj)
+		if err != nil {
+			c.logger.Errorf("Error adding CronJob trigger controller as finalizer to  CronJobTrigger Obj: %s CRD object due to: %v: ", key, err)
+			return err
+		}
+	}
+
+	or, err := utils.GetCronJobTriggerOwnerReference(cronJobtriggerObj)
+	if err != nil {
+		return err
+	}
+
+	restIface := c.clientset.BatchV2alpha1().RESTClient()
+	groupVersion, err := c.getResouceGroupVersion("cronjobs")
```

Comment:
as far as I remember this was added to support CronJobs both in 1.7 and 1.8 because the client-go version at that time didn't allow to deploy CronJobs for both versions. I don't know if it is still the case.

Reply:
.. isn't that an excellent example of why we _should_ check for a few known/explicit groupVersions here?  If the server is wildly different to the structure that `utils.EnsureCronJob` knows how to create, then we might be creating something completely invalid.  In that case "Server doesn't support cronjobs version v1beta.whatever" is going to be a _much_ clearer error message for the user than "your cronjob subtly misbehaved or complained about an obscure invalid schema value".